### PR TITLE
[Snappi] Adding Multidut Global Pause RDMA Testcase

### PR DESCRIPTION
**Pull request by dks0692**
### Description of PR

- Tests global pause traffic in a multidut/multi-asic  environment.

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205

#### What is the motivation for this PR?
To verify the global pause behavior in multi-dut/multi-asic scenario.
#### How did you verify/test it?
Tested against voq chassis with SONiC installed in it
#### Any platform specific information?
Tested against J2C+
#### Supported testbed topology if it's a new test case?
T2, T1 & T0
### Documentation
https://github.com/sonic-net/sonic-mgmt/blob/master/docs/testplan/pfc/GLOBAL_PAUSE_README.md
### Output

johnar@24b40a6e217b:~/sonic-mgmt/tests$ py.test --inventory ../ansible/snappi-sonic --host-pattern sonic-s6100-dut --testbed vms-snappi-sonic --testbed_file ../ansible/testbed.csv --show-capture=stdout --log-cli-level info --showlocals -ra --allow_recover --skip_sanity --disable_loganalyzer snappi/pfc/test_multidut_global_pause_with_snappi.py 
/usr/local/lib/python2.7/dist-packages/ansible/parsing/vault/__init__.py:44: CryptographyDeprecationWarning: Python 2 is no longer supported by the Python core team. Support for it is now deprecated in cryptography, and will be removed in the next release.
  from cryptography.exceptions import InvalidSignature
================================================================================= test session starts ==================================================================================
platform linux2 -- Python 2.7.17, pytest-4.6.5, py-1.10.0, pluggy-0.13.1
ansible: 2.8.12
rootdir: /var/johnar/sonic-mgmt/tests, inifile: pytest.ini
plugins: forked-1.3.0, html-1.22.1, metadata-1.11.0, repeat-0.9.1, xdist-1.28.0, ansible-2.2.2
collecting ... 
--------------------------------------------------------------------------------- live log collection ----------------------------------------------------------------------------------
14:39:09 __init__.pytest_collection_modifyitems   L0541 INFO   | Available basic facts that can be used in conditional skip:
{
  "topo_type": "ptf", 
  "testbed": "vms-snappi-sonic", 
  "topo_name": "ptf64"
}

------------------------------------------------------------------------------------ live log setup ------------------------------------------------------------------------------------
14:39:09 __init__.set_default                     L0049 INFO   | Completeness level not set during test execution. Setting to default level: CompletenessLevel.basic
14:39:09 __init__.check_test_completeness         L0138 INFO   | Test has no defined levels. Continue without test completeness checks
14:39:17 conftest.generate_params_dut_hostname    L0937 INFO   | Using DUTs ['sonic-s6100-dut'] in testbed 'vms-snappi-sonic'
14:39:17 conftest.rand_one_dut_hostname           L0291 INFO   | Randomly select dut sonic-s6100-dut for testing
14:39:17 conftest.creds_on_dut                    L0562 INFO   | dut sonic-s6100-dut belongs to groups [u'snappi-sonic', u'sonic', u'sonic_dell64_40', 'fanout']
14:39:17 conftest.creds_on_dut                    L0585 INFO   | skip empty var file ../ansible/group_vars/all/corefile_uploader.yml
14:39:17 conftest.creds_on_dut                    L0585 INFO   | skip empty var file ../ansible/group_vars/all/env.yml
14:39:20 conftest.nbrhosts                        L0421 INFO   | No VMs exist for this topology: ptf64
14:39:20 __init__.sanity_check                    L0125 INFO   | Skip sanity check according to command line argument
14:39:20 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture enable_packet_aging_after_test setup starts --------------------
14:39:20 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture enable_packet_aging_after_test setup ends --------------------
14:39:20 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture rand_lossless_prio setup starts --------------------
14:39:20 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture rand_lossless_prio setup ends --------------------
14:39:20 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture rand_lossy_prio setup starts --------------------
14:39:20 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture rand_lossy_prio setup ends --------------------
14:39:20 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture start_pfcwd_after_test setup starts --------------------
14:39:20 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture start_pfcwd_after_test setup ends --------------------
14:39:20 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture snappi_api_serv_ip setup starts --------------------
14:39:20 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture snappi_api_serv_ip setup ends --------------------
14:39:20 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture snappi_api_serv_port setup starts --------------------
14:39:20 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture snappi_api_serv_port setup ends --------------------
14:39:20 __init__._fixture_generator_decorator    L0081 INFO   | -------------------- fixture snappi_api setup starts --------------------
14:39:20 __init__._fixture_generator_decorator    L0085 INFO   | -------------------- fixture snappi_api setup ends --------------------
14:39:20 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture prio_dscp_map setup starts --------------------
14:39:22 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture prio_dscp_map setup ends --------------------
14:39:22 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture lossless_prio_list setup starts --------------------
14:39:25 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture lossless_prio_list setup ends --------------------
14:39:25 __init__.loganalyzer                     L0045 INFO   | Log analyzer is disabled
14:39:25 __init__._fixture_func_decorator         L0069 INFO   | -------------------- fixture get_multidut_snappi_ports setup starts --------------------
14:39:25 __init__._fixture_func_decorator         L0076 INFO   | -------------------- fixture get_multidut_snappi_ports setup ends --------------------
------------------------------------------------------------------------------------ live log call -------------------------------------------------------------------------------------
14:39:25 snappi_fixtures.__intf_config_multidut   L0728 INFO   | Configuring Dut: sonic-s6100-dut with port Ethernet68 with IP 20.0.1.1/24
14:39:27 snappi_fixtures.__intf_config_multidut   L0728 INFO   | Configuring Dut: sonic-s6100-dut with port Ethernet64 with IP 20.0.2.1/24
14:39:43 connection._warn                         L0332 WARNING| Verification of certificates is disabled
14:39:43 connection._info                         L0329 INFO   | Determining the platform and rest_port using the 10.36.77.63 address...
14:39:43 connection._warn                         L0332 WARNING| Unable to connect to http://10.36.77.63:11009.
14:39:43 connection._info                         L0329 INFO   | Connection established to `https://10.36.77.63:11009 on windows`
14:39:43 connection._info                         L0329 INFO   | Using IxNetwork api server version 9.30.2212.5
14:39:43 connection._info                         L0329 INFO   | User info IxNetwork/ST-REG-VM/8010
14:39:43 snappi_api.info                          L1132 INFO   | snappi-0.9.1
14:39:43 snappi_api.info                          L1132 INFO   | snappi_ixnetwork-0.8.2
14:39:43 snappi_api.info                          L1132 INFO   | ixnetwork_restpy-1.1.10
14:39:43 snappi_api.info                          L1132 INFO   | Config validation 0.012s
14:39:44 snappi_api.info                          L1132 INFO   | Ports configuration 0.057s
14:39:44 snappi_api.info                          L1132 INFO   | Captures configuration 0.030s
14:39:44 snappi_api.info                          L1132 INFO   | Location hosts ready [10.36.78.53] 0.018s
14:39:44 snappi_api.info                          L1132 INFO   | Speed conversion is not require for (port.name, speed) : [('Port 0', u'novusHundredGigNonFanOut'), ('Port 1', u'novusHundredGigNonFanOut')]
14:39:44 snappi_api.info                          L1132 INFO   | Aggregation mode speed change 0.243s
14:39:44 snappi_api.info                          L1132 INFO   | Location preemption [10.36.78.53;4;1, 10.36.78.53;4;2] 0.018s
14:40:08 snappi_api.info                          L1132 INFO   | Location connect [Port 0, Port 1] 24.100s
14:40:08 snappi_api.info                          L1132 INFO   | Location state check [Port 0, Port 1] 0.042s
14:40:08 snappi_api.info                          L1132 INFO   | Location configuration 24.554s
14:40:17 snappi_api.info                          L1132 INFO   | Layer1 configuration 9.001s
14:40:17 snappi_api.info                          L1132 INFO   | Lag Configuration 0.009s
14:40:17 snappi_api.info                          L1132 INFO   | Convert device config : 0.069s
14:40:17 snappi_api.info                          L1132 INFO   | Create IxNetwork device config : 0.000s
14:40:18 snappi_api.info                          L1132 INFO   | Push IxNetwork device config : 0.222s
14:40:18 snappi_api.info                          L1132 INFO   | Devices configuration 0.300s
14:40:21 snappi_api.info                          L1132 INFO   | Flows configuration 3.371s
14:40:26 snappi_api.info                          L1132 INFO   | Start interfaces 4.974s
14:40:26 snappi_api.info                          L1132 INFO   | IxNet - The Traffic Item was modified. Please perform a Traffic Generate to update the associated traffic Flow Groups
14:40:26 traffic_generation.run_traffic           L0285 INFO   | Wait for Arp to Resolve ...
14:40:26 traffic_generation.run_traffic           L0296 INFO   | Starting transmit on all flows ...
14:40:28 snappi_api.info                          L1132 INFO   | Flows generate/apply 1.793s
14:40:39 snappi_api.info                          L1132 INFO   | Flows clear statistics 10.601s
14:40:39 snappi_api.info                          L1132 INFO   | Captures start 0.000s
14:40:43 snappi_api.info                          L1132 INFO   | Flows start 3.640s
14:40:48 traffic_generation.run_traffic           L0337 INFO   | Stop transmit on all flows ...
14:40:54 snappi_api.info                          L1132 INFO   | Flows stop 5.504s
14:40:54 snappi_fixtures.cleanup_config           L0881 INFO   | Removing Configuration on Dut: sonic-s6100-dut with port Ethernet68 with ip :20.0.1.1/24
14:40:56 snappi_fixtures.cleanup_config           L0881 INFO   | Removing Configuration on Dut: sonic-s6100-dut with port Ethernet64 with ip :20.0.2.1/24
PASSED                                                                                                                                                                           [100%]
---------------------------------------------------------------------------------- live log teardown -----------------------------------------------------------------------------------
14:40:57 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture snappi_api teardown starts --------------------
14:40:57 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture snappi_api teardown ends --------------------
14:40:57 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture start_pfcwd_after_test teardown starts --------------------
14:40:59 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture start_pfcwd_after_test teardown ends --------------------
14:40:59 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture rand_lossy_prio teardown starts --------------------
14:40:59 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture rand_lossy_prio teardown ends --------------------
14:40:59 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture rand_lossless_prio teardown starts --------------------
14:40:59 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture rand_lossless_prio teardown ends --------------------
14:40:59 __init__._fixture_generator_decorator    L0093 INFO   | -------------------- fixture enable_packet_aging_after_test teardown starts --------------------
14:41:04 __init__._fixture_generator_decorator    L0102 INFO   | -------------------- fixture enable_packet_aging_after_test teardown ends --------------------

=================================================================================== warnings summary ===================================================================================
/usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:538
  /usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:538: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: tests.common.plugins.loganalyzer
    self.import_plugin(import_spec)

/usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:538
  /usr/local/lib/python2.7/dist-packages/_pytest/config/__init__.py:538: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: tests.common.dualtor
    self.import_plugin(import_spec)

snappi/pfc/test_multidut_global_pause_with_snappi.py::test_global_pause[linecard_configuration_set0-non_chassis_single_line_card]
  /usr/local/lib/python2.7/dist-packages/pytest_ansible-2.2.2-py2.7.egg/pytest_ansible/module_dispatcher/v28.py:70: UserWarning: provided hosts list is empty, only localhost is available
    warnings.warn("provided hosts list is empty, only localhost is available")

-- Docs: https://docs.pytest.org/en/latest/warnings.html
-------------------------------------------------------------------------------- live log sessionfinish --------------------------------------------------------------------------------
14:41:05 __init__.pytest_terminal_summary         L0064 INFO   | Can not get Allure report URL. Please check logs
==================== 1 passed, 3 warnings in 117.16 seconds ====================
INFO:root:Can not get Allure report URL. Please check logs
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?

### CODE REVIEW

**Comment by wangxin** [comment]
@dks0692 The pre-commit check is mandatory now. PR merging is blocked if it failed. Can you follow the pre-commit local checking steps to ensure that the change can pass pre-commit check? Basically, the issue is that file `tests/common/snappi/qos_fixtures.py` needs an empty line at the end of file.

File: tests/common/snappi/qos_fixtures.py [hunk 1]
```diff
@@ -25,7 +25,7 @@ def prio_dscp_map(duthosts, rand_one_dut_hostname):
         Example: {0: [0], 1: [1], 2: [2], 3: [3], 4: [4] ....}
     """
     duthost = duthosts[rand_one_dut_hostname]
-    config_facts = duthost.config_facts(host=duthost.hostname,
+    config_facts = duthost.config_facts(host=duthost.hostname, asic_index=0,
```
Comment (kenneth-arista) [inline tests/common/snappi/qos_fixtures.py:28]: Why is the asic namespace hardcoded to 0?

Comment (dks0692) [reply tests/common/snappi/qos_fixtures.py:28]: this was the fixture written for multi-dut infra , to take in first hostname

File: tests/snappi/pfc/files/multidut_helper.py [hunk 2]
```diff
@@ -0,0 +1,435 @@
+import time
```
Comment (baiwei0427) [inline tests/snappi/pfc/files/multidut_helper.py:1]: We've recently implemented updates to the Snappi tests, relocating all test cases to the following location: https://github.com/sonic-net/sonic-mgmt/tree/master/tests/snappi_tests. Kindly make the necessary adjustments to reflect this change.

Furthermore, I kindly request you take a look at the pull request at https://github.com/sonic-net/sonic-mgmt/pull/8660, contributed by @developfast . Please ensure that the PFC tests for multi DUT environments align harmoniously with the existing PFC tests that have been refactored by @developfast .

**Review by baiwei0427 — CHANGES_REQUESTED** [review]

**Comment by developfast** [comment]
Please add the appropriate new parameters added to this PR to `snappi_test_params`

File: tests/snappi/pfc/files/multidut_helper.py (hunk 2 above)
Comment (selldinesh) [reply tests/snappi/pfc/files/multidut_helper.py]: this has been fixed

File: tests/common/snappi_tests/qos_fixtures.py [hunk 3]
```diff
@@ -25,17 +25,17 @@ def prio_dscp_map(duthosts, rand_one_dut_hostname):
         Example: {0: [0], 1: [1], 2: [2], 3: [3], 4: [4] ....}
     """
     duthost = duthosts[rand_one_dut_hostname]
-    config_facts = duthost.config_facts(host=duthost.hostname, asic_index=0,
+    config_facts = duthost.config_facts(host=duthost.hostname,
                                         source="running")['ansible_facts']
 
-    if "DSCP_TO_TC_MAP" not in list(config_facts.keys()):
+    if "DSCP_TO_TC_MAP" not in config_facts.keys():
```
Comment (developfast) [inline tests/common/snappi_tests/qos_fixtures.py:31]: why are we reverting back to just keys()

File: tests/common/snappi_tests/qos_fixtures.py [hunk 4]
```diff
@@ -73,18 +73,18 @@ def lossless_prio_list(duthosts, rand_one_dut_hostname):
         Lossless priorities (list)
     """
     duthost = duthosts[rand_one_dut_hostname]
-    config_facts = duthost.config_facts(host=duthost.hostname, asic_index=0,
+    config_facts = duthost.config_facts(host=duthost.hostname,
```
Comment (developfast) [inline tests/common/snappi_tests/qos_fixtures.py:76]: do we not need an asic index?

File: tests/common/snappi_tests/snappi_test_params.py [hunk 5]
```diff
@@ -22,7 +22,21 @@ def __init__(self):
                                              then pcap is on the tgen egress port
             base_flow_config (dict): base flow configuration
             test_tx_frames (list): number of test frames transmitted for priorities to test ex. [2000, 3000]
-                for priorities 3 and 4
+                                    for priorities 3 and 4
+            duthost1 (obj): DUT 1 in case of multi-dut testbed environment
+            rx_port (dict): Contains the details of the rx port
```
Comment (developfast) [inline tests/common/snappi_tests/snappi_test_params.py:27]: This is already in the `base_flow_config` parameter of this object - see `base_flow_config["rx_port_config"]` or `base_flow_config["tx_port_config"]`

**Review by developfast — CHANGES_REQUESTED** [review]
some repetitive code that needs to be cleaned up

File: tests/common/snappi_tests/snappi_test_params.py (hunk 5 above)
Comment (developfast) [reply tests/common/snappi_tests/snappi_test_params.py:27]: You can look at the `setup_base_flow_config()` function in `traffic_generation.py`

Comment (dks0692) [reply tests/common/snappi_tests/snappi_test_params.py]: fixed this

File: tests/common/snappi_tests/qos_fixtures.py (hunk 4 above)
Comment (dks0692) [reply tests/common/snappi_tests/qos_fixtures.py]: fixed this

File: tests/common/snappi_tests/qos_fixtures.py (hunk 3 above)
Comment (dks0692) [reply tests/common/snappi_tests/qos_fixtures.py]: fixed this

**Review by developfast — CHANGES_REQUESTED** [review]

File: tests/common/snappi_tests/snappi_test_params.py [hunk 6]
```diff
@@ -33,3 +43,12 @@ def __init__(self):
...
         self.test_tx_frames = 0
+        self.duthost1 = None
+        self.duthost2 = None
+        self.multidut_ports = None
+        self.kmin = 0
+        self.kmax = 0
+        self.pmax = 0
+        self.pkt_size = 0
+        self.pkt_cnt = 0
+        self.iters = 1
```
[the remark this answers is not in the document]
Comment (dks0692) [reply tests/common/snappi_tests/snappi_test_params.py]: this has been resolved

File: tests/common/snappi_tests/snappi_test_params.py [hunk 7]
```diff
@@ -33,3 +43,12 @@ def __init__(self):
         self.is_snappi_ingress_port_cap = True
         self.base_flow_config = None
         self.test_tx_frames = 0
+        self.duthost1 = None
+        self.duthost2 = None
+        self.multidut_ports = None
+        self.kmin = 0
```
[the remark this answers is not in the document]
Comment (dks0692) [reply tests/common/snappi_tests/snappi_test_params.py]: this has been resolved

File: tests/common/snappi_tests/snappi_test_params.py [hunk 8]
```diff
@@ -33,3 +43,12 @@ def __init__(self):
...
+        self.kmax = 0
+        self.pmax = 0
+        self.pkt_size = 0
```
[the remark this answers is not in the document]
Comment (dks0692) [reply tests/common/snappi_tests/snappi_test_params.py]: yeah i can make this as global variable

File: tests/snappi_tests/multidut/pfc/test_multidut_global_pause_with_snappi.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        duthosts (pytest fixture): list of DUTs
+        lossless_prio_list (pytest fixture): list of all the lossless priorities
+        prio_dscp_map (pytest fixture): priority vs. DSCP map (key = priority).
+        line_card_choice: Line card choice to be mentioned in the variable.py file
+        linecard_configuration_set : Line card classification, (min 1 or max 2  hostnames and asics to be given)
+    Returns:
+        N/A
+    """
+    if line_card_choice not in linecard_configuration_set.keys():
+        assert False, "Invalid line_card_choice value passed in parameter"
```
[the remark this answers is not in the document]
Comment (dks0692) [reply tests/snappi_tests/multidut/pfc/test_multidut_global_pause_with_snappi.py]: fixed this

**Review by developfast — CHANGES_REQUESTED** [review]

File: tests/common/snappi_tests/qos_fixtures.py [hunk 10]
```diff
@@ -106,3 +106,64 @@ def lossy_prio_list(all_prio_list, lossless_prio_list):
     """
     result = [x for x in all_prio_list if x not in lossless_prio_list]
     return result
+
+
+def prio_dscp_map_dut_base(duthost):
```
[the remark this answers is not in the document]
Comment (dks0692) [reply tests/common/snappi_tests/qos_fixtures.py]: reverted the file back to the contents of main, as this fixture is not needed

File: tests/common/snappi_tests/snappi_test_params.py [hunk 11]
```diff
@@ -33,3 +36,5 @@ def __init__(self):
         self.is_snappi_ingress_port_cap = True
         self.base_flow_config = None
         self.test_tx_frames = 0
+        self.multi_dut_params = MultiDUTParams()
+        self.test_iterations = 1
```
[the remark this answers is not in the document]
Comment (dks0692) [reply tests/common/snappi_tests/snappi_test_params.py:40]: used for ecn cases

File: tests/snappi_tests/multidut/pfc/files/multidut_helper.py [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+BG_FLOW_NAME = 'Background Flow'
+BG_FLOW_AGGR_RATE_PERCENT = 45
+DATA_PKT_SIZE = 1024
+data_flow_pkt_size = 1024
+DATA_FLOW_DURATION_SEC = 2
+DATA_FLOW_DELAY_SEC = 1
+data_flow_delay_sec = 1
+SNAPPI_POLL_DELAY_SEC = 2
+TOLERANCE_THRESHOLD = 0.05
+PORT_SPEED = 'speed_400_gbps'
```
[the remark this answers is not in the document]
Comment (dks0692) [reply tests/snappi_tests/multidut/pfc/files/multidut_helper.py]: this variable is not used in the script, removed them

**Review by developfast — CHANGES_REQUESTED** [review]

**Review by developfast — APPROVED** [review]

**Review by developfast — APPROVED** [review]
approved

**Pull request merged**